src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionDiagnosticAnalyzer.AnalysisResult.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertSwitchStatementToExpression
+{
+    internal sealed partial class ConvertSwitchStatementToExpressionDiagnosticAnalyzer
+    {
+        private enum AnalysisResultKind
+        {
+            // Note: default(AnalysisResult) should yield a
+            // failure if we ever reached that in the visitor.
+            Failure = 0,
+
+            // Note: we will use this for either a "throw" statement
+            // or as the insignificant part of a combination, thus we
+            // named it "Neutral" rather than "Throw" to avoid confusion.
+            Neutral,
+
+            Break,
+            Return,
+            Assignment,
+        }
+
+        private readonly struct AnalysisResult
+        {
+            private readonly AnalysisResultKind _kind;
+
+            // Holds assignment target expressions in each switch
+            // section to ensure that they are all the same.
+            private readonly ImmutableArray<ExpressionSyntax> _assignmentTargets;
+
+            private AnalysisResult(AnalysisResultKind kind, ImmutableArray<ExpressionSyntax> assignmentTargets = default)
+            {
+                _kind = kind;
+                _assignmentTargets = assignmentTargets;
+            }
+
+            public static AnalysisResult Failure => new AnalysisResult(AnalysisResultKind.Failure);
+            public static AnalysisResult Return => new AnalysisResult(AnalysisResultKind.Return);
+            public static AnalysisResult Break => new AnalysisResult(AnalysisResultKind.Break);
+            public static AnalysisResult Neutral => new AnalysisResult(AnalysisResultKind.Neutral);
+            public static AnalysisResult Assignment(ImmutableArray<ExpressionSyntax> targets) => new AnalysisResult(AnalysisResultKind.Assignment, targets);
+            public static AnalysisResult Assignment(ExpressionSyntax target) => new AnalysisResult(AnalysisResultKind.Assignment, ImmutableArray.Create(target));
+
+            public bool IsFailure => _kind == AnalysisResultKind.Failure;
+            public bool IsBreak => _kind == AnalysisResultKind.Break;
+            public bool IsReturn => _kind == AnalysisResultKind.Return;
+            public bool IsNeutral => _kind == AnalysisResultKind.Neutral;
+            public bool IsAssignment => _kind == AnalysisResultKind.Assignment;
+
+            public bool Success => IsReturn || IsAssignment;
+
+            public override string ToString() => _kind.ToString();
+
+            public static AnalysisResult Combine(AnalysisResult left, AnalysisResult right)
```

Comment:
>Combine [](start = 41, length = 7)

Regarding naming: The way I understand it, `Combine` is for combining statements within a section, and `Match` is for intersecting the results across the whole `switch` (multiple sections).
Maybe one is "Union" and the other is "Intersect"?

Should those be instance methods?

Reply:
Oooh, i like that :)